File: Arrays/unique-3-digit-even-numbers/ts/unique3DigitEven.ts
export {};

function totalNumbers(digits: number[]): number {
  //   let unique = new Set<number>(); not as optimized as bool array
  let vis = Array.from({ length: 1000 }, () => false);
  let count = 0;
  for (let i = 0; i <= digits.length; i++) {
    if (digits[i] == 0) {
      continue;
    }
    for (let j = 0; j <= digits.length; j++) {
      for (let k = 0; k <= digits.length; k++) {
        if (i == j || i == k || k == j) {
          continue;
        }
        let digit = digits[i] * 100 + digits[j] * 10 + digits[k];
        if (digit % 2 == 0 && !vis[digit]) {
          vis[digit] = true;
          count++;
          // unique.add(digit)
        }
      }
    }
  }
  return count;
}

const testCases: number[][] = [
  [1, 2, 3],
  [2, 2, 8, 8, 2],
  [0, 2, 0],
  [0, 0, 0],
  [1, 3, 5],
  [0, 1, 2, 3],
];

for (const digits of testCases) {
  console.log(`Input:       ${JSON.stringify(digits)}`);
  console.log(`Output:      ${totalNumbers(digits)}`);
  console.log("-".repeat(35));
}

File: Arrays/unique-3-digit-even-numbers/go/unique_3digit_even.go
package main

import (
	"fmt"
	"strings"
)

func Contains[T comparable](slice []T, target T) bool {
	for _, item := range slice {
		if item == target {
			return true
		}
	}
	return false
}

func totalNumbers(digits []int) int {
	count := 0
	vis := make([]bool, 1000)
	for i := range digits {
		for j := range digits {
			for k := range digits {
				if i == j || i == k || k == j {
					continue
				}
				digit := digits[i]*100 + digits[j]*10 + digits[k]
				//I previously used Contains on a separate slice to check if already seen
				if digit%2 == 0 && digit/100 > 0 && !vis[digit] {
					vis[digit] = true
					count++
				}
			}
		}
	}
	return count
}

func main() {
	testCases := [][]int{
		{1, 2, 3},
		{2, 2, 8, 8, 2},
		{0, 2, 0},
		{0, 0, 0},
		{1, 3, 5},
		{0, 1, 2, 3},
	}

	for _, digits := range testCases {
		fmt.Printf("Input:       %v\n", digits)
		fmt.Printf("Output:      %d\n", totalNumbers(digits))
		fmt.Println(strings.Repeat("-", 35))
	}
}
